server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -90,20 +91,30 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
                     throw new IllegalStateException("cannot add duplicate task: " + a);
                 }, IdentityHashMap::new));
             LinkedHashSet<BatchedTask> newTasks = new LinkedHashSet<>(tasks);
-            tasksPerBatchingKey.merge(firstTask.batchingKey, newTasks, (existingTasks, updatedTasks) -> {
-                for (BatchedTask existing : existingTasks) {
+            // Need to maintain below order in which task identity map and task map are updated.
+            // For insert: First insert identity in taskIdentity map with dup check and then insert task in taskMap.
+            // For remove: First remove task from taskMap and then remove identity from taskIdentity map.
+            // We are inserting identity first and removing at last to ensure no duplicate tasks are enqueued.
+            // Changing this order might lead to duplicate tasks in queue.
+            taskIdentityPerBatchingKey.merge(firstTask.batchingKey, tasksIdentity, (existingIdentities, updatedIdentities) -> {
+                for (Object newIdentity : tasksIdentity.keySet()) {
```

Comment:
We should use local variable(method argument value - existingIdentities) inside the remappingFunction implementation rather than using global variable (**taskIdentity** is a global variable for remapping function).